### ISSUE TITLE
unable to start scsynth for non-QT build

### ISSUE THREAD

**Issue by rhetr** [supercollider/supercollider#2882]
I just compiled supercollider for ubuntu 16.04 from the latest git with these flags:
```
cmake -DSC_QT=OFF -DNO_AVAHI=ON -DSC_EL=no ..
```
I compiled and installed without issue. running `scsynth -v` yields
```
scsynth 3.9dev (Built from branch 'master' [2a105f1])
```

However, I'm unable to successfully start the server. `s` returns `nil`, so I have to manually set it to `s = Server.default`. Then running 
```
s = Server.default;
s.boot;
```
i get
```
ERROR: Message 'put' not understood.
RECEIVER:
   nil
ARGS:
   Integer 26565
Instance of Function {    (0x22788e8, gc=20, fmt=00, flg=00, set=02)
  instance variables [2]
    def : instance of FunctionDef in Method Server:bootServerApp
    context : Frame (0x2279618) of Server:bootServerApp
}
CALL STACK:
	DoesNotUnderstandError:reportError   0x2278f38
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError   0x2279058
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError   0x2279108
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError   0x22791b8
		arg this = <instance of Routine>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw   0x2279268
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand   0x227a088
		arg this = nil
		arg selector = 'put'
		arg args = [*2]
	String:unixCmd   0x227abc8
		arg this = " -u 57110 -a 116 -i 2 -o 2 -..."
		arg action = <instance of Function>
		arg postOutput = true
		var pid = 26565
	Server:bootServerApp   0x2279618
		arg this = <instance of Server>
		arg onComplete = <instance of Function>
	Routine:prStart   0x227ae28
		arg this = <instance of Routine>
		arg inval = 11.619275113
^^ The preceding error dump is for ERROR: Message 'put' not understood.
RECEIVER: nil

/bin/sh: 1: -u: not found
RESULT = 127
ERROR: Message 'at' not understood.
RECEIVER:
   nil
ARGS:
   Integer 26565
CALL STACK:
	DoesNotUnderstandError:reportError   0x22774d8
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError   0x2277588
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError   0x22776a8
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw   0x2277758
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand   0x22785f8
		arg this = nil
		arg selector = 'at'
		arg args = [*1]
	Meta_String:doUnixCmdAction   0x2278718
		arg this = <instance of Meta_String>
		arg res = 127
		arg pid = 26565
^^ The preceding error dump is for ERROR: Message 'at' not understood.
RECEIVER: nil
```

I specified `Server.program` manually by adding `Server.program = '/usr/local/bin/scsynth';` which gives me
```
-> localhost
ERROR: Message 'put' not understood.
RECEIVER:
   nil
ARGS:
   Integer 26399
Instance of Function {    (0x22ab0b8, gc=70, fmt=00, flg=00, set=02)
  instance variables [2]
    def : instance of FunctionDef in Method Server:bootServerApp
    context : Frame (0x21ac048) of Server:bootServerApp
}
CALL STACK:
	DoesNotUnderstandError:reportError   0x2276568
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError   0x233d4d8
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError   0x22b9718
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError   0x22f0458
		arg this = <instance of Routine>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw   0x222f0a8
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand   0x22ad7f8
		arg this = nil
		arg selector = 'put'
		arg args = [*2]
	String:unixCmd   0x223d468
		arg this = "/usr/local/bin/scsynth -u 57..."
		arg action = <instance of Function>
		arg postOutput = true
		var pid = 26399
	Server:bootServerApp   0x21ac048
		arg this = <instance of Server>
		arg onComplete = <instance of Function>
	Routine:prStart   0x22b0558
		arg this = <instance of Routine>
		arg inval = 492.108043339
^^ The preceding error dump is for ERROR: Message 'put' not understood.
RECEIVER: nil


JackDriver: client name is 'SuperCollider'
SC_AudioDriver: sample rate = 44100.000000, driver's block size = 128
SuperCollider 3 server ready.
```
This starts scsynth but I'm unable to connect to the instance, as `s.serverRunning` yields `false` and running `{ SinOsc.ar * 0.1 }.play(s);` gets me
```
WARNING: server 'localhost' not running.
```

**Comment by patrickdupuis** [comment]
What are your motivations in building SC `NO_AVAHI=ON` and `SC_QT=OFF`?

**Comment by patrickdupuis** [comment]
There are currently issues with building without Qt (see #2144). As for Avahi, I believe it is essential for SC to function properly.

**Comment by jamshark70** [comment]
The reason for the error is that String's classvar unixCmdActions is not initialized -- it should be a dictionary, but it's nil in your case.

The variable should be initialized in `initClass`, so this means class library initialization is not completing successfully.

I recently saw problems with class library init when trying a no-QT build, so I'll take a wild guess that you're seeing something similar. In that case, there would be other errors in the startup sequence which you haven't shared here.

**Comment by rhetr** [comment]
oh, yeah there were some QT-related errors but I just assumed they didn't matter. here:

```
init_OSC
compiling class library...
initPassOne started
	NumPrimitives = 591
initPassOne done
	compiling dir: '/usr/local/share/SuperCollider/SCClassLibrary'
	compiling dir: '/usr/local/share/SuperCollider/Extensions'
	compiling dir: '/home/rhetr/.local/share/SuperCollider/Extensions'
	pass 1 done
	numentries = 810563 / 11232210 = 0.072
	5055 method selectors, 2222 classes
	method table size 12239928 bytes, big table size 89857680
	Number of Symbols 11641
	Byte Code Size 353149
	compiled 316 files in 0.95 seconds

Info: 4 methods are currently overwritten by extensions. To see which, execute:
MethodOverride.printAll

compile done
ERROR: A primitive was not bound. 0 590
Instance of Method {    (0x2009680, gc=01, fmt=00, flg=11, set=04)
  instance variables [15]
    raw1 : Float 0.000000   00000000 0070000C
    raw2 : Float 0.000000   00000200 02010002
    code : instance of Int8Array (0x2009980, size=4, set=2)
    selectors : nil
    constants : nil
    prototypeFrame : instance of Array (0x20098c0, size=2, set=2)
    context : nil
    argNames : instance of SymbolArray (0x2009800, size=2, set=2)
    varNames : nil
    sourceCode : nil
    ownerClass : class Meta_QtGUI (0x1fc5240)
    name : Symbol 'prSetStyle'
    primitiveName : Symbol '_Qt_SetStyle'
    filenameSymbol : Symbol '/usr/local/share/SuperCollider/SCClassLibrary/Common/GUI/Base/QtGUI.sc'
    charPos : Integer 2872
}
ERROR: Primitive '__none' failed.
Failed.
RECEIVER:
   nil

PROTECTED CALL STACK:
	Meta_MethodError:new	0x1497740
		arg this = PrimitiveFailedError
		arg what = Failed.
		arg receiver = nil
	Meta_PrimitiveFailedError:new	0x149dd00
		arg this = PrimitiveFailedError
		arg receiver = nil
	Object:primitiveFailed	0x10da2c0
		arg this = nil
	Function:prTry	0x16ea080
		arg this = a Function
		var result = nil
		var thread = a Thread
		var next = nil
		var wasInProtectedFunc = false
	
CALL STACK:
	MethodError:reportError   0x11f2678
		arg this = <instance of PrimitiveFailedError>
	Nil:handleError   0x11f25c8
		arg this = nil
		arg error = <instance of PrimitiveFailedError>
	Thread:handleError   0x11f2518
		arg this = <instance of Thread>
		arg error = <instance of PrimitiveFailedError>
	Object:throw   0x11f2468
		arg this = <instance of PrimitiveFailedError>
	Function:protect   0x11ee0a8
		arg this = <instance of Function>
		arg handler = <instance of Function>
		var result = <instance of PrimitiveFailedError>
	Meta_QtGUI:style_   0x11edf48
		arg this = <instance of Meta_QtGUI>
		arg styleName = "Fusion"
	Meta_QtGUI:initClass   0x2346598
		arg this = <instance of Meta_QtGUI>
	Meta_Class:initClassTree   0x23343e8
		arg this = <instance of Meta_Class>
		arg aClass = <instance of Meta_QtGUI>
		var implementsInitClass = nil
	ArrayedCollection:do   0x23308b8
		arg this = [*203]
		arg function = <instance of Function>
		var i = 55
	Meta_Class:initClassTree   0x232fab8
		arg this = <instance of Meta_Class>
		arg aClass = <instance of Meta_Object>
		var implementsInitClass = nil
	Process:startup   0x232eb38
		arg this = <instance of Main>
		var time = 0.967011975
	Main:startup   0x232d548
		arg this = <instance of Main>
		var didWarnOverwrite = false
^^ The preceding error dump is for ERROR: Primitive '__none' failed.
Failed.
RECEIVER: nil
```

according to the README both QT5 and Avahi are optional, I didn't have either installed and was trying to get supercollider compiled quickly. I'll install those libs and recompile but I don't really see why they should be requirements..

**Comment by jamshark70** [comment]
Right, so what is happening is, a GUI method fails (because the backend doesn't support it), and this interrupts the initialization process. Anything that was supposed to initialize after that will not be set up, so you may get all kinds of strange failures.

When it's working, you should see a line like this in the startup sequence:

```
Class tree inited in 0.08 seconds
```

That GUI method should ideally not be called at all in a non-GUI installation. But, the build system doesn't include logic to exclude the GUI parts of the class library (which is annoying). But you can do that by hand:

https://github.com/supercollider/supercollider/issues/2144#issuecomment-302536579

> if you build with -DSC_IDE=OFF and -DSC_QT=OFF, the trick to get the library to compile is to...
> ```
> mv /usr/local/share/SuperCollider/SCClassLibrary/Common/GUI /usr/local/share/SuperCollider/SCClassLibrary/scide_scqt/GUI
> mv /usr/local/share/SuperCollider/SCClassLibrary/JITLib/GUI /usr/local/share/SuperCollider/SCClassLibrary/scide_scqt/JITLibGUI
> ```
> 
> long standing issue

**Comment by jamshark70** [comment]
Oh... for that matter, issue #2144 is about a different problem with non-qt builds -- quite serious one. So the best solution is to install Qt 5.whatever.

AFAIK this problem has nothing to do with Avahi (but I think I've always built with Avahi, so I can't say if there are or are not other problems).

**Comment by vivid-synth** [comment]
I always build with ```-DNO_AVAHI=ON``` and have had no issues.

**Comment by vivid-synth** [comment]
If you'd rather not build/install QT and are ok with sclang wasting CPU in the short term (till #2144 if fixed), the lines above may be enough to make it work:

```
mv /usr/local/share/SuperCollider/SCClassLibrary/Common/GUI /usr/local/share/SuperCollider/SCClassLibrary/scide_scqt/GUI
mv /usr/local/share/SuperCollider/SCClassLibrary/JITLib/GUI /usr/local/share/SuperCollider/SCClassLibrary/scide_scqt/JITLibGUI
```

**Comment by rhetr** [comment]
Thanks for all the info, i built with qt and both with/without avahi and they are both working fine. I'll leave this thread open as it is still an issue though for non QT builds.

**Comment by rhetr** [comment]
Thanks for all the info, i built with qt and both with/without avahi and they are both working fine. I'll leave this thread open as it is still an issue though for non QT builds.

**Comment by snappizz** [comment]
if i'm understanding this issue correctly, i think it can be subsumed into #3778.

**Comment by JordanHendersonMusic** [comment]
Closing as I can't reproduce, seems to be fixed by one of prs mentioned in #3778
